weed/filer/mysql_store/mysql_store.go

```diff
@@ -0,0 +1,224 @@
+package mysql_store
+
+import (
+	"database/sql"
+	"fmt"
+	"hash/crc32"
+	"sync"
+	"time"
+
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const (
+	sqlUrl             = "%s:%s@tcp(%s:%d)/%s?charset=utf8"
+	maxIdleConnections = 100
+	maxOpenConnections = 50
+	maxTableNums       = 1024
+	tableName          = "filer_mapping"
+)
+
+var (
+	_init_db        sync.Once
+	_db_connections []*sql.DB
+)
+
+type MySqlConf struct {
+	User     string
+	Password string
+	HostName string
+	Port     int
+	DataBase string
+}
+
+type MySqlStore struct {
+	dbs []*sql.DB
+}
+
+func getDbConnection(confs []MySqlConf) []*sql.DB {
+	_init_db.Do(func() {
+		for _, conf := range confs {
+
+			sqlUrl := fmt.Sprintf(sqlUrl, conf.User, conf.Password, conf.HostName, conf.Port, conf.DataBase)
+			var dbErr error
+			_db_connection, dbErr := sql.Open("mysql", sqlUrl)
+			if dbErr != nil {
+				_db_connection.Close()
+				_db_connection = nil
+				panic(dbErr)
+			}
+			_db_connection.SetMaxIdleConns(maxIdleConnections)
+			_db_connection.SetMaxOpenConns(maxOpenConnections)
+			_db_connections = append(_db_connections, _db_connection)
+		}
+	})
+	return _db_connections
+}
+
+func NewMysqlStore(confs []MySqlConf) *MySqlStore {
+	ms := &MySqlStore{
+		dbs: getDbConnection(confs),
+	}
+
+	for _, db := range ms.dbs {
+		for i := 0; i < maxTableNums; i++ {
+			if err := ms.createTables(db, tableName, i); err != nil {
+				fmt.Printf("create table failed %s", err.Error())
+			}
+		}
+	}
+
+	return ms
+}
+
+func (s *MySqlStore) hash(fullFileName string) (instance_offset, table_postfix int) {
+	hash_value := crc32.ChecksumIEEE([]byte(fullFileName))
+	instance_offset = int(hash_value) % len(s.dbs)
+	table_postfix = int(hash_value) % maxTableNums
+	return
+}
+
+func (s *MySqlStore) parseFilerMappingInfo(path string) (instanceId int, tableFullName string, err error) {
+	instance_offset, table_postfix := s.hash(path)
+	instanceId = instance_offset
+	tableFullName = fmt.Sprintf("%s_%04d", tableName, table_postfix)
+	return
+}
+
+func (s *MySqlStore) Get(fullFilePath string) (fid string, err error) {
+	instance_offset, tableFullName, err := s.parseFilerMappingInfo(fullFilePath)
+	if err != nil {
+		return "", err
+	}
+	fid, err = s.query(fullFilePath, s.dbs[instance_offset], tableFullName)
+	if err == sql.ErrNoRows {
+		//Could not found
+		err = nil
```

Comment:
To follow Golang's convention, any error should not be simply passed to the upper level.

For mysql specific error, it should be wrapped to another error meaningful to current logic. Just return the error will work, but not recommended.

if err == mysql.NotFound {
    return fmt.Errorf("Failed to find key: %v", err)
}